Fix usage of Netty pooled allocator with Virtual Threads

franz1981 opened the pull request:
This PR contains 2 changes:
- **Fix NettyLoomAdaptorProcessor due to Netty's PoolThreadCache changes**: this is fixing the change in the constructor at https://github.com/netty/netty/pull/13510
- **Allocate heap buffers within virtual threads** (which fixes #39061 and alleviate #39696 despite not being reported :"( - which seems a OpenJDK bug to investigate, really)

@cescoffier TBH I would get rid of `NettyLoomAdaptorProcessor`, but there could be other parts in Quarkus/Vertx or extensions which *could* allocate pooled buffer from virtual threads, and in that case, despite hitting #39061, it is still better than the original code behaviour.
If we feel to remove it instead, I'm all in to do it in this same PR, saving the first fix.

franz1981 committed:
Fix NettyLoomAdaptorProcessor due to Netty's PoolThreadCache changes

franz1981 committed:
Allocate heap buffers within virtual threads

franz1981 commented:
I still have to run some benchmark, but the virtual thread CI job doesn't work (last time it has run was +1 yr ago!) - let me know if you prefer me to make it draft.

@cescoffier do you want me to remove the netty processor or is fine as it is?

cescoffier commented:
Let's make it a draft.

About the processor, I'm not totally sure we still need it.

franz1981 commented on independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java:
```diff
@@ -1,53 +1,81 @@
 final class AppendBuffer {
-    private final ByteBufAllocator allocator;
+
+    private static final MethodHandle virtualMh = PlatformDependent.javaVersion() >= 21 ? findVirtualMH() : null;
```
maybe we have a better way which work with both JVM and native image here @geoand ? IDK

geoand replied:
We'll probably need a GraalVM substitution for this

franz1981 replied:
Consider that vertx does the same exact thing at https://github.com/eclipse-vertx/vert.x/blob/adbe97600cc6215f15ce2fac629c89f93618ca8f/src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java#L37 but without checking the JDK version (that's not needed I think) - which means that *if* we use the HybridJacksonPool it's going to do the same as I do here.

geoand replied:
It might even work OOTB in the latest GraalVM versions, I am not sure really

franz1981 commented:
@cescoffier 

Yep, if the numbers will be decent with this changes, I suppose we can get rid of a tons of code, till Netty won't provide a more user-friendly (and NOT broken, from a virtual thread pov) buffer pool.

What we could do as well, to fix this, wold be to reuse the created virtual threads, basically sharing a (bounded) task queue with them and just looping to consume tasks from there. Clearly this is not a nice solution but will make thread local(s) to work - with a HUGE performance impact in case they are meant to store heavyweight objects, given that the resident footprint will now be multiplied but the huge number of v thread(s) created (and parked).

franz1981 commented:
My concern here @cescoffier is related https://github.com/quarkusio/quarkus/blob/c51f136003962457b13f716c5eae75dd37d53669/extensions/netty-loom-adaptor/deployment/src/main/java/io/quarkus/netty/loom/adaptor/NettyLoomAdaptorProcessor.java#L543

which is used in the hot path, regardless being a virtual thread: it shows that JDK 21 tests can have some relevant performance impact because of https://openjdk.org/jeps/416 which says

> When Method, Constructor, and Field instances are held in non-constant fields (e.g., in a non-final field or an array element), microbenchmarks show some performance degradation. The performance of field accesses is significantly slower than the old implementation, by 51–77%, when Field instances cannot be constant-folded.

Sadly `isVirtual` will be in the hot path, meaning that will slow down any threads trying to use Netty's pooled allocator(s).

I see 2 solutions for this:
- in this PR remove the Netty processor 
- leave it but make all the reflective accesses static final (and some `MethodHandle`, too, see https://github.com/quarkusio/quarkus/issues/33102#issuecomment-2022853621)

franz1981 committed:
Remove NettyLoomAdaptorProcessor

franz1981 commented:
@cescoffier 
@geoand 
I've added a commit to remove the processor: the TLDR story is:
- when we use JDK 21 and `Thread::isVirtual` and `Thread::currentCarrierThread` are avaialble, `canUseVirtual` is true (see https://github.com/quarkusio/quarkus/blob/67f06de033b258ff46cb0653fa50ebf9a8ca5eb0/extensions/netty-loom-adaptor/deployment/src/main/java/io/quarkus/netty/loom/adaptor/NettyLoomAdaptorProcessor.java#L232)
- BUT, due to a change on Netty, fixed by 73e8bcc5b77a75d17256ed7e35e5fb3efac72efe, the processor *always* fail to work (ie is broken)
- this is VERY lucky for us, because https://github.com/quarkusio/quarkus/pull/39704#issuecomment-2023117557 explain that if it would be fixed, will cause a regression on **all** the Netty allocations (which happen pretty much everywhere) due to the reflective access to `Thread::isVirtual` happening in the very hot path.

This is the reason why I've removed the processor: thanks to another fix we don't need it anymore, and we will save to fix these other problems which will happen, by adapting correctly to the new Netty API.

In short, we've been lucky it was broken!!!

franz1981 commented:
Let's wait the CI, but right now, at least for Jackson, this seems to work as expected.

@geoand I would like to use the information collected at build-time for the specific end point where the `AppendBuffer` is used, to understand *if* it can run on a virtual thread or just to avoid checking for virtual threads using reflection, if not necessary....

geoand commented:
> @geoand I would like to use the information collected at build-time for the specific end point where the AppendBuffer is used, to understand if it can run on a virtual thread or just to avoid checking for virtual threads using reflection, if not necessary....

So IIUC what you are looking for is to know whether the user's REST method (assuming one was invoked) was annotated with `@RunOnVirtualThread`?

franz1981 commented:
yep @geoand , so I don't need to check if the Thread is virtual and I could use a holder pattern to avoid looking up eagerly to the method handle (eg using a VirtualChecker class which is used only *if* the `AppendBuffer` is configured to expect virtual thread(s) executors to use it.

geoand commented:
The thing is that it's not really safe to do it any other way because the execution model of the REST method is not guaranteed to be the same for every subsequent piece of code.

geoand commented:
The only thins safe to assume is that if the REST was **not** executed on a virtual thread, then neither will any of the subsequent pieces of code. 
Is that a strong enough assumption for you to use?

franz1981 commented:
@geoand 

> Is that a strong enough assumption for you to use?

At this point is fine as it is, I think

The change I'm pushing is basically checking that whoever is the appender's usage context (which depends on the stream) - it's going to NOT use pooled Netty buffer(s) from virtual thread(s), because the pooled allocator of Netty is not yet compatible with them. I believe it was (from the beginning, when the processor was created) what we wanted: it wasn't possible to do it before, because the allocation logic was coupled with the stream, while now we have the append buffer, which guarantee some flexibility and try hard to not overallocate (unless configured to do it).

franz1981 commented:
The only thing, which concern me, is that IDK if the Netty composite buffer can contains buffer(s) of different types (unpooled heap based and pooled direct ones) - and *if* that can ever happen.

> the execution model of the REST method is not guaranteed to be the same for every subsequent piece of code.

This one: I don't know if users can append writes (the stream could be used directly by users too AFAIK) from different types of thread(s) and how the composite buffer would behave (as said earlier) - I'm writing a Netty test to verify this

geoand commented:
> I don't know if users can append writes (the stream could be used directly by users too AFAIK)

Users don't know about this layer, but they can certainly provide an `InputStream` which represents the data that is to be written out

franz1981 commented:
Got it, so, as it is now should work, but I still want to make sure Netty is happy for the composite buffer (just in case)

franz1981 commented:
Netty seems happy with this usage or the composite buffer, so @geoand , just need to check if native image is fine with my MethodHandle.
I am deleting a lot of code with this PR too, but @cescoffier is fine, so, ready to go let's say :)

geoand commented:
Have you rebased onto `main` lately? 

I would like to get an up to date run from CI :)

franz1981 commented:
Just rebased now @geoand let's see. We need a JDK 21 run to make it uses the code path I have added

geoand commented:
You mean for JDK or native image?

franz1981 commented:
Actually both.
- JDK 21 because will trigger the need to check if a thread is virtual or not 
- native image to make sure that both MethodHandle and checking the platform version work reliably

geoand commented:
For JDK we use 17, 21 so we are covered. 

For native image we use `quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-21`

franz1981 commented:
@geoand https://github.com/quarkusio/quarkus/actions/runs/8508575040/job/23302917533?pr=39704 this seems the only error, but I cannot say if it's something related my pr...

geoand commented:
Right, totally unrelated

franz1981 commented on independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java:
```diff
@@ -1,53 +1,81 @@
+            return false;
+        }
+        try {
+            return (boolean) virtualMh.invokeExact(Thread.currentThread());
```
@galderz @zakkak can we verify that this one is as fast as a direct call in native-image?
I want to make sure we don't pay for this given that is in the fastest-path

galderz replied:
@franz1981 I'll look into it.
